akka-persistence-typed/src/test/scala/docs/akka/persistence/typed/AccountExample6b.scala

```diff
@@ -0,0 +1,90 @@
+import akka.persistence.typed.scaladsl.PersistentBehaviors.CommandHandler
+import akka.persistence.typed.scaladsl.PersistentBehaviors6.HandlerFactory
+import akka.persistence.typed.scaladsl.{Effect, PersistentBehaviors6}
+import docs.akka.persistence.typed.AccountExample6b.AccountEntity.{Account, AccountCommand, AccountEvent}
+
+object AccountExample6b {
+
+  object AccountEntity extends HandlerFactory[AccountCommand,  AccountEvent, Account]{
+
+    sealed trait AccountCommand
+    case object CreateAccount extends AccountCommand
+    case class Deposit(amount: Double) extends AccountCommand
+    case class Withdraw(amount: Double) extends AccountCommand
+    case object CloseAccount extends AccountCommand
+
+    sealed trait AccountEvent
+    case object AccountCreated extends AccountEvent
+    case class Deposited(amount: Double) extends AccountEvent
+    case class Withdrawn(amount: Double) extends AccountEvent
+    case object AccountClosed extends AccountEvent
+
+
+    sealed trait Account {
+      def applyCommand: CommandHandler
+      def applyEvent: EventHandler
+    }
+
+    case object EmptyAccount extends Account {
+
+      override val applyCommand = CommandHandler.partial {
+        case CreateAccount ⇒ Effect.persist(AccountCreated)
+      }
+
+      override val applyEvent = EventHandler.partial {
+        case AccountCreated ⇒ OpenedAccount(0.0)
+      }
+
+    }
+
+    case class OpenedAccount(balance: Double) extends Account {
+
+      override val applyCommand = CommandHandler.partial {
+          case Deposit(amount) ⇒ Effect.persist(Deposited(amount))
+
+          case Withdraw(amount) ⇒
+            if ((this.balance - amount) < 0.0)
+              Effect.unhandled // TODO replies are missing in this example
+            else {
+              Effect
+                .persist(Withdrawn(amount))
+                .thenRun {
+                  case OpenedAccount(balance) ⇒
+                    // do some side-effect using balance
+                    println(balance)
+                  case _ ⇒ throw new IllegalStateException
+                }
+            }
+          case CloseAccount if this.balance == 0.0 ⇒
+            Effect.persist(AccountClosed)
+        }
+
+
+      override val applyEvent = EventHandler.partial {
+        case Deposited(amount) ⇒ copy(balance = balance + amount)
+        case Withdrawn(amount) ⇒ copy(balance = balance - amount)
+        case AccountClosed     ⇒ ClosedAccount
+      }
+
+    }
+
+    case object ClosedAccount extends Account {
+      override val applyCommand = CommandHandler.unhandled
+      override val applyEvent = EventHandler.unhandled
+    }
+
+
+    def behavior(accountNumber: String): Behavior[AccountEntity.AccountCommand] = {
+      PersistentBehaviors6.receive[AccountCommand,  AccountEvent, Account](
+        accountNumber,
+        EmptyAccount,
+        (state, cmd) => state.applyCommand(cmd),
+        (state, event) ⇒ state.applyEvent(event)
+      )
+    }
```

Comment:
I see the potential of such factories to reduce boilerplate and support certain cases. It's great that the core API is minimalistic and these are only helpers.

I find [AccountExampe6](https://github.com/akka/akka/pull/25503/files#diff-8c562db98b836446fd204c15507596ea) more straightforward (you immediately see what it is because it's only explicit methods and functions).

When documenting these things I think we should have a gentle introduction by first showing plain functions and then explain that the same thing can be done with these helpers to reduce the boilerplate. We probably need a whole chapter in the docs explaining different styles.

We should explore what we can do in the javadsl.

Reply:
My personal opinion: While I like the idea in this PR, I don't think it pulls its weight when it comes to introducing more API surface—and as @patriknw says: it's essentially abstracting away two functions.

Reply:
The main reason for this is to remove boilerplate. There is also the introduction of `PartialFunction` that are converted to total functions. Not a must, but a convenience. 

In [AccountExample6](https://github.com/akka/akka/pull/25503/files#diff-8c562db98b836446fd204c15507596ea) we removed the need for repeating `CommandHandler[AccountCommand, AccountEvent, Account]]` over and over again. 

This was done it two fronts. First with a type alias.
```
type Effect = akka.persistence.typed.scaladsl.Effect[AccountEvent, Account]
```

Secondly, by defining:

```scala
sealed trait Account {
  def applyCommand(cmd: AccountCommand): Effect
  def applyEvent(event: AccountEvent): Account
}
```

`CommandHandler[Command, Event, State]` in the current API is an alias to `(State, Command) ⇒ Effect[Event, State]`.

So, with `type Effect` alias, it becomes `(State, Command) => Effect` and with the `applyCommand` in the Account trait, it becomes `Command => Effect`.

There are many ways for removing that kind of boilerplate. We can say, create a type alias or extend this helper class.

Reply:
Another advantage of factories is that we can provide some pre-define boilerplate removal facilities and I find it better than documenting how to remove boilerplate. Users can just pick the style they want.